### ISSUE TITLE
"ComputeCallSignature failed" in verbose logs investigating performance loss.

### ISSUE THREAD

**Issue by cjchristopher** [jax-ml/jax#30270]
I've been trying to track down what I perceive as a massive performance loss after upgrading to JAX 0.6.0 (and subsequently 0.6.1 and 0.6.2) as well as a a few Nvidia/CUDA driver updates. This has ultimately led me to enable maximum logging, and making some changes to jaxlib/pjit.cc (printing `function_name_` for all the early `return` paths around and after [`ComputeCallSignature`](https://github.com/jax-ml/jax/blob/8ff5d3345feb9b34e191becabb477defdb7a5ace/jaxlib/pjit.cc#L668) alongside the failure messages)

I am calling vmap over a function that calls a [jaxopt L-BFGS-B (`.run())`](https://github.com/google/jaxopt/blob/b12c2f3ddaa3ef71fac42d7e419adccb9dcc49a1/jaxopt/_src/lbfgsb.py#L227) (I'm aware jaxopt is now archived but withstanding another JAX-based L-BFGS-B implementation, I'm stuck with it. The issue occurs with other jaxopt solvers as well - but I think this is more fundamental than jaxopt?) - not sure whether my actual objective function is required here, but can provide details if necessary.

```python
solver = jaxopt.LBFGSB(fun=my_eval)

def opt(x, weights):
    bounds = (-1 * jax.numpy.ones_like(x), jax.numpy.ones_like(x))
    x_opt, state = self.solver.run(x, weights=weights, bounds=bounds)
    return x_opt, state

vect_opt = jax.vmap(opt, in_axes=(0,None))
res = vect_opt(xs, weights)
```

Here's what I'm getting my logs (truncated):
```
I jaxlib/pjit.cc:668] ComputeCallSignature failed: INVALID_ARGUMENT: Not supported: The C++ ToPyArgSignature only accepts Buffer/DeviceArray, Numpy arrays scalars of supported types (see implementation), or Python scalars. Got type <class 'jax._src.interpreters.batching.BatchTracer'>
For function _where
---
I jaxlib/pjit.cc:668] ComputeCallSignature failed: INVALID_ARGUMENT: Not supported: The C++ ToPyArgSignature only accepts Buffer/DeviceArray, Numpy arrays scalars of supported types (see implementation), or Python scalars. Got type <class 'jax._src.interpreters.batching.BatchTracer'>
For function _where
I jaxlib/pjit.cc:668] ComputeCallSignature failed: INVALID_ARGUMENT: Not supported: The C++ ToPyArgSignature only accepts Buffer/DeviceArray, Numpy arrays scalars of supported types (see implementation), or Python scalars. Got type <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
For function _broadcast_arrays
---
```

and this one which seem like its the entirety of the L-BFGS-B loop (truncating the INVALID_ARGUMENT string from above):
```
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...batching.BatchTracer'> For function update
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...ad.JVPTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...ad.JVPTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...ad.JVPTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...ad.JVPTracer'> For function _reduce_sum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function fista_line_search
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function negative
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function clip
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function maximum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function minimum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function subtract
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function real
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function square
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function imag
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function square
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function _reduce_sum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function evaluate
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function _reduce_sum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function subtract
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function vdot
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function ravel
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function ravel
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function dot
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function greater
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function less_equal
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function logical_and
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function negative
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function clip
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function less_equal
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function true_divide
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function _where
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function _broadcast_arrays
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function sqrt
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function subtract
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function subtract
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function true_divide
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function multiply
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function real
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function square
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function imag
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function square
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function _reduce_sum
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function sqrt
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function true_divide
I jaxlib/pjit.cc:668] ComputeCallSignature failed: ... Got type <class '...partial_eval.DynamicJaxprTracer'> For function add`
```
as well as others. More than a few `I jaxlib/pjit.cc:739] cpp pjit fallback to python.` hanging around in the logs as well.

It's unclear to me from poking around the jaxlib source whether this is expected behaviour or not - hoping someone can shed some light on this one, or at least tell me this is normal.

### System info (python version, jaxlib version, accelerator, etc.)

```
$ python
Python 3.12.7 (main, Jun 18 2025, 13:16:51) [GCC 14.2.0] on linux
>>> import jax; jax.print_environment_info()
jax:    0.7.0.dev20250716+8ff5d3345
jaxlib: 0.7.0.dev20250717
numpy:  2.3.1
python: 3.12.7 (main, Jun 18 2025, 13:16:51) [GCC 14.2.0]
device info: NVIDIA RTX 2000 Ada Generation Laptop GPU-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='peregrine', release='6.11.0-29-generic', version='#29-Ubuntu SMP PREEMPT_DYNAMIC Fri Jun 13 20:29:41 UTC 2025', machine='x86_64')

$ nvidia-smi
Thu Jul 17 15:00:57 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.158.01             Driver Version: 570.158.01     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA RTX 2000 Ada Gene...    Off |   00000000:01:00.0 Off |                  N/A |
| N/A   35C    P8              3W /   35W |     120MiB /   8188MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A           73991      G   /usr/bin/gnome-shell                      2MiB |
|    0   N/A  N/A         3500367      C   python                                  100MiB |
+-----------------------------------------------------------------------------------------+
```
(I'll note this also happens with 0.6.2 installed via pip following the normal install guide)

**Comment by cjchristopher** [comment]
So I've been restructuring my code and it seems that simply calling `jax.jit()` over the `vmap`'d function call resolves the issue - which is perhaps to be expected. It would appear this is in part due to the `jit=bool` parameter that most of the `jaxopt` solvers use, which is `True` by default and so the full optimisation loop attempts compilation, but due to the outer vmap, we are obviously passing around Tracers to an underlying `jit`ted function.

I would be half tempted to make a note of this on jaxopt as a "sharp bits" if it weren't archived, but I guess this pitfall might apply to many routines provided packages in the JAX ecosystem that one might wish to vectorise/shard over if those things do any `jit` under the hood?

**Comment by cjchristopher** [comment]
Actually I need to reopen this. After further testing, `jit`ing the outer vmap appears to not solve the problem. In fact, applying or not applying `jax.jit` variously up and down my call stack (manually on my evaluation function, with `jit=False` or `jit=True` given to the solver, inside the vmap, outside the vmap, all or none of the above etc) I still observe this behaviour.

Since calling a `vmap`ped or `jit`ted function seems to necessarily make the inputs in to tracer types, jaxlib being unable to compute the CallSignatures and falling back to python seems to be quite the problem - how have I managed to mess this up?

Edit: I note these comments here regarding permissible types - https://github.com/jax-ml/jax/blob/8ff5d3345feb9b34e191becabb477defdb7a5ace/jaxlib/pjit.cc#L641-L657 - I've tried explicitly committing my input arrays to my single GPU before calling the compiled function and still no joy - there's not much of any mention of requiring commitment for compilation in the documentation when `jit` or `vmap` or composing those two is mentioned anywhere, but in any case it didn't seem to fix the problem

**Comment by vvvm23** [comment]
I am seeing a similar issue, also from moving to jax `0.6.x`

```
ComputeCallSignature failed: INVALID_ARGUMENT: Not supported: The C++ ToPyArgSignature only accepts Buffer/DeviceArray, Numpy arrays scalars of supported types (see implementation), or Python scalars. Got type <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
```

@cjchristopher did you manage to get anywhere with this?

**Comment by cjchristopher** [comment]
> [@cjchristopher](https://github.com/cjchristopher) did you manage to get anywhere with this?

@vvvm23 As best I can tell the old advice used to be that you could `jit` just about everything, but with recent updates and refactors to both JAX and XLA I think the actual best practice is to just `jit` at the outer most point in your code - I elaborate on this a bit here https://github.com/jax-ml/jax/discussions/10104#discussioncomment-13841959 but I haven't heard back from anyone yet.

As far as I can tell these messages are intended and not indicative of a problem - you can reduce them somewhat by removing unnecessary `jit` calls but you'll still get them since half of the built-in JAX routines are automatically compiled, as are say, some methods in jaxopt (if you are using it). The entire call chain gets compiled for your outer most compilation directive but the inner `jit` annotations are still seen and so compilation is attempted "in the midst" of the current compilation (hence why it sees Tracers), and then bails and returns to the other compilation if was already doing.

My performance loss ended up being explained by other issues with my code, mostly to do with accidentally mixing different sharding approaches in maximally bad ways (and slightly also by what I also believe are bugs in JAXOpt).
